### PR TITLE
Start Ray on the head and the worker nodes

### PR DESCRIPTION
**Pull request by DenisYay**
The algorithm:
* We connect to the existing Ray instance started in `_start_server_cmds` and initialized inside `http_server.py`. The instance listens on port 6379 (note: the Skypilot ray instance listens on port 6380).
* We start the required number of Ray workers as part of the restart login in `cluster.py`. The workers start on the number of nodes that is derived from cluster.ips (-1 for the head node)
* We await for all workers to join the Ray cluster in `http_server.py` (verified by a call to `ray.nodes()` )
* Once the above completes successfully, we proceed with the restart flow, otherwise an error is thrown

### CODE REVIEW

File: tests/fixtures/on_demand_cluster_fixtures.py [hunk 1]
```diff
@@ -79,13 +79,17 @@ def ondemand_https_cluster_with_auth():
...
+    }
     c = rh.ondemand_cluster(**args)
     init_args[id(c)] = args
 
     c.up_if_not()
 
     c.save()
 
-    c.install_packages(["pytest"])
+    c.install_packages(["pytest", "ray"])
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply tests/fixtures/on_demand_cluster_fixtures.py:94]: Done

File: tests/test_resources/test_clusters/test_multinode_cluster.py [hunk 2]
```diff
@@ -27,3 +27,14 @@ def test_rsync_and_ssh_onto_worker_node(self, multinode_cpu_cluster):
...
 
         assert "runhouse" in status_codes[0][1]
+
+    @pytest.mark.level("thorough")
+    def test_ray_started_on_worker_node_after_cluster_restart(
+        self, multinode_cpu_cluster
+    ):
+        worker_node = multinode_cpu_cluster.ips[-1]
+
+        multinode_cpu_cluster.restart_server(restart_ray=True)
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply tests/test_resources/test_clusters/test_multinode_cluster.py:37]: Agree, removed

File: runhouse/resources/hardware/cluster.py [hunk 3]
```diff
@@ -698,6 +703,54 @@ def _start_server_cmds(
 
         return cmds
 
+    def _start_ray(self, host, master_host, n_hosts, ray_port):
+        if host == master_host:
+            # Head node
+            if ray.is_initialized():
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py]: Removed the current check.

We could run
`nmap -sV --reason -p 6381 127.0.0.1`

and check that we get

PORT     STATE SERVICE REASON         VERSION
6381/tcp open  redis?  syn-ack

Problem: we need to install nmap.

or

`nc -vv -z 127.0.0.1 6381`

Connection to domain_name 6381 port [tcp/*] succeeded!

But this result is too coarse imho (not Ray specific).

Please let me know if you have a preference between the two, but I don't think it's a blocker for a first merge.

File: runhouse/resources/hardware/cluster.py [hunk 4]
```diff
@@ -698,6 +703,54 @@ def _start_server_cmds(
...
+                logger.info(
+                    f"There is a Ray cluster already running on the head node {master_host}. Shutting it down."
+                )
+                ray.shutdown()
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py]: Yep, removed the coarse code.

File: runhouse/resources/hardware/cluster.py [hunk 5]
```diff
@@ -787,6 +840,17 @@ def restart_server(
...
+            # Restart Ray on the head node and each of the workers
+            # TODO: kill ray on all nodes first. Need to think more of the
+            # multiple multi-node clusters running on the same set of machines case.
+            master_host = self.address
+            n_hosts = len(self.ips)
+            for host in self.ips:
+                self._start_ray(host, master_host, n_hosts, self.DEFAULT_RAY_PORT)
+
+            # logger.info("🎉 All workers present and accounted for 🎉")
+            # logger.info(ray.cluster_resources())
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py]: +1

File: runhouse/resources/hardware/cluster.py [hunk 6]
```diff
@@ -698,6 +703,54 @@ def _start_server_cmds(
...
+            worker_thread = threading.Thread(
+                target=self._wait_for_workers, args=(self, n_hosts)
+            )
+            worker_thread.start()
+
+        else:
+            # Worker node
+            self.run(
+                commands=[
+                    f"sleep 10 && ray start --address={master_host}:{ray_port} --block",
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py]: did you mean the on the worker nodes? I'm not sure that ray start supports that? `ray.init(address="ray://123.45.67.89:6381", namespace="runhouse")` does, but I'm not sure that that's the recommended way to connect to the cluster: https://docs.ray.io/en/latest/ray-core/configure.html#cluster-resources

File: runhouse/resources/hardware/cluster.py [hunk 7]
```diff
@@ -787,6 +840,17 @@ def restart_server(
...
             self.client.cert_path = self.cert_config.cert_path
 
+        if restart_ray:
+            # Restart Ray on the head node and each of the workers
+            # TODO: kill ray on all nodes first. Need to think more of the
+            # multiple multi-node clusters running on the same set of machines case.
+            master_host = self.address
+            n_hosts = len(self.ips)
+            for host in self.ips:
+                self._start_ray(host, master_host, n_hosts, self.DEFAULT_RAY_PORT)
```
[the remark this answers is not in the document]
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py]: Yes, I've updated the code to start remotely on the head node and remotely on the workers.

Without the explicit start on the head node, we have two ray servers: 6380 by SkyPilot and 6379 default.

Do you mean we should use the existing 6379 default one and connect the workers to it?

[the remark this answers is not in the document]
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py]: As based on an offline convo, we are reusing the existing Ray instance on port 6379

File: runhouse/resources/hardware/cluster.py [hunk 8]
```diff
@@ -698,6 +699,33 @@ def _start_server_cmds(
...
+        public_head_ip = master_host
+
+        # Find the internal IP corresponding to the public_head_ip
+        internal_head_ip = None
+        for internal, external in self.live_state["handle"][
+            "stable_internal_external_ips"
+        ]:
+            if external == public_head_ip:
+                internal_head_ip = internal
+                break
```
Comment (dongreenberg) [inline runhouse/resources/hardware/cluster.py:713]: Nice!

File: runhouse/resources/hardware/cluster.py [hunk 9]
```diff
@@ -787,6 +815,15 @@ def restart_server(
             self.client.use_https = https_flag
             self.client.cert_path = self.cert_config.cert_path
 
+        if restart_ray:
+            # Restart Ray on the head node and each of the workers
+            # TODO: kill ray on all nodes first. Need to think more of the
```
Comment (dongreenberg) [inline runhouse/resources/hardware/cluster.py:820]: We should find out if it dies automatically on the workers when we kill it on the head or if we're really borking them when we pkill the head node.

File: runhouse/resources/hardware/cluster_factory.py [hunk 10]
```diff
@@ -307,6 +307,10 @@ def ondemand_cluster(
         else open_ports
     )
 
+    # Add the new Runhouse Ray port to open_ports if it's not already there
+    if Cluster.DEFAULT_RAY_PORT not in open_ports:
+        open_ports.append(Cluster.DEFAULT_RAY_PORT)
+
```
Comment (dongreenberg) [inline runhouse/resources/hardware/cluster_factory.py:313]: 😬 makes me nervous. If we use internal IPs do we really need to open it? Wouldn't this expose the cluster to the outside world?

File: tests/test_resources/test_clusters/test_multinode_cluster.py [hunk 11]
```diff
@@ -27,3 +27,12 @@ def test_rsync_and_ssh_onto_worker_node(self, multinode_cpu_cluster):
...
         assert "runhouse" in status_codes[0][1]
+
+    @pytest.mark.level("thorough")
+    def test_ray_started_on_worker_node_after_cluster_restart(
+        self, multinode_cpu_cluster
+    ):
+        head_node = multinode_cpu_cluster.ips[0]
+
+        status_codes = multinode_cpu_cluster.run(["ray status"], node=head_node)
+        assert status_codes[0][0] == 0
```
Comment (dongreenberg) [inline tests/test_resources/test_clusters/test_multinode_cluster.py:38]: We should check that two nodes show up in the output, right?

File: tests/fixtures/on_demand_cluster_fixtures.py [hunk 12]
```diff
@@ -79,7 +79,11 @@ def ondemand_https_cluster_with_auth():
 
 @pytest.fixture(scope="session")
 def multinode_cpu_cluster():
-    args = {"name": "rh-cpu-multinode", "num_instances": 2, "instance_type": "CPU:2+"}
+    args = {
+        "name": "rh-cpu-multinode-777",
```
Comment (dongreenberg) [inline tests/fixtures/on_demand_cluster_fixtures.py:83]: Is 777 lucky?

Comment (DenisYay) [reply tests/fixtures/on_demand_cluster_fixtures.py]: It was! But it's working now, so seems like we don't need luck any more ;)

File: runhouse/resources/hardware/cluster_factory.py (hunk 10 above)
Comment (DenisYay) [reply runhouse/resources/hardware/cluster_factory.py:313]: Works without, removed.

File: runhouse/resources/hardware/cluster.py (hunk 9 above)
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py:820]: Seems like 'RAY_HEAD_IP=127.0.0.1 RAY_HEAD_PORT=6379 ray stop' should stop the entire cluster with workers when run on the head node.

File: tests/test_resources/test_clusters/test_multinode_cluster.py (hunk 11 above)
Comment (DenisYay) [reply tests/test_resources/test_clusters/test_multinode_cluster.py:38]: Does .run() return the console output?

File: runhouse/resources/hardware/cluster.py (hunk 9 above)
Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py:820]: Nope, it doesn't seem to work..

Back to pkill for now. 

The workers do not die automatically, they become orphaned. There seems to be a way to carefully send shutdown commands to them, but need to experiment further. We can probably decouple it from the current PR?

Comment (DenisYay) [reply runhouse/resources/hardware/cluster.py:820]: Next to try: python3 -m "import ray; ray.init('ray://localhost:6379'); ray.shutdown()"
https://docs.ray.io/en/latest/ray-core/api/doc/ray.shutdown.html

If that doesn't work, maybe this is a more involved and explicit way:
https://stackoverflow.com/questions/69613739/how-to-kill-ray-tasks-when-the-driver-is-dead

**Pull request merged**